extensions/mysql/lib/mysql.go

```diff
@@ -0,0 +1,439 @@
+	"io"
+	"io/ioutil"
+	"os"
+	"os/user"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+)
+
+type Uploader struct {
+	*internal.Uploader
+}
+type Backup struct {
+	*internal.Backup
+}
+
+type ArchiveNonExistenceError internal.ArchiveNonExistenceError
+
+func HandleStreamFetch(backupName string, folder internal.StorageFolder) {
+	if backupName == "" || backupName == "LATEST" {
+		latest, err := internal.GetLatestBackupName(folder)
+		if err != nil {
+			tracelog.ErrorLogger.Fatalf("Unable to get latest backup %+v\n", err)
+		}
+		backupName = latest
+	}
+	stat, _ := os.Stdout.Stat()
+	if (stat.Mode() & os.ModeCharDevice) == 0 {
+	} else {
+		tracelog.ErrorLogger.Fatalf("stout is a terminal")
+	}
+	err := downloadAndDecompressStream(folder, backupName)
+	if err != nil {
+		tracelog.ErrorLogger.Fatalf("%+v\n", err)
+	}
+}
+
+// TODO : unit tests
+func downloadAndDecompressStream(folder internal.StorageFolder, fileName string) error {
+	baseBackupFolder := folder.GetSubFolder(internal.BaseBackupPath)
+	backup := Backup{internal.NewBackup(baseBackupFolder, fileName)}
+
+	streamSentinel, err := backup.fetchStreamSentinel()
+	if err != nil {
+		return err
+	}
+	binlogsAreDone := make(chan error)
+
+	go fetchBinlogs(folder, streamSentinel, binlogsAreDone)
+
+	for _, decompressor := range internal.Decompressors {
+		d := decompressor
+		archiveReader, exists, err := internal.TryDownloadWALFile(baseBackupFolder, getStreamName(&backup, d.FileExtension()))
+		if err != nil {
+			return err
+		}
+		if !exists {
+			continue
+		}
+		err = internal.DecompressWALFile(&internal.EmptyWriteIgnorer{WriteCloser: os.Stdout}, archiveReader, d)
+		os.Stdout.Close()
+
+		tracelog.DebugLogger.Println("Waiting for binlogs")
+		err = <-binlogsAreDone
+
+		return err
+	}
+	return internal.NewArchiveNonExistenceError(fmt.Sprintf("Archive '%s' does not exist.\n", fileName))
+}
+
+func fetchBinlogs(folder internal.StorageFolder, sentinel StreamSentinelDto, binlogsAreDone chan error) {
+	binlogFolder := folder.GetSubFolder(BinlogPath)
+	endTS, dstFolder := GetBinlogConfigs()
+	if dstFolder == "" {
+		binlogsAreDone <- errors.New("WALG_MYSQL_BINLOG_DST is not configured")
+		return
+	}
+	objects, _, err := binlogFolder.ListFolder()
+	if err != nil {
+		binlogsAreDone <- nil
+		return
+	}
+
+	for _, object := range objects {
+		tracelog.DebugLogger.Println("Consider binlog ", object.GetName(), object.GetLastModified().Format(time.RFC3339))
+
+		binlogName := ExtractBinlogName(object, folder)
+
+		if BinlogShouldBeFetched(sentinel, binlogName, endTS, object) {
+			fileName := path.Join(dstFolder, binlogName)
+			tracelog.DebugLogger.Println("Download", binlogName, "to", fileName)
+			err := internal.DownloadWALFileTo(binlogFolder, binlogName, fileName)
+			if err != nil {
+				binlogsAreDone <- err
+				return
+			}
+		}
+	}
+
+	binlogsAreDone <- nil
+}
+
+func CheckOldBinlogs(backupTime internal.BackupTime, folder internal.StorageFolder) {
```

Comment:
This function not only Checks old binlogs, but also deletes them...

Reply:
Renamed to DeleteOldBinlogs, because it seems to be ok if it wouldn't be found